### ISSUE TITLE
Changing the font size in individual panels

### ISSUE THREAD

**Issue by meurigbeynon** [EMGroup/js-eden#136]
At one point, we had +/- buttons to enlarge the font in the 'Input Window'. This was an important feature when presenting construals, as it wasn't appropriate to scale the entire display, but useful to be able to make what was being input visible to an audience. I think having this facility attached to other display components is equally important - it's hard for anyone to see what is displayed in an Observable List, for instance. Would be good to see this feature restored / extended id possible,

**Comment by knicos** [comment]
Easy to do using css3 transform (transform: translate(50%,50%); scale(2.0,2.0)). Where do we want these buttons? Are they always available? Or would an _view observable be wanted? And should all input windows scale together?

**Comment by knicos** [comment]
${{ $('.outputcontent, .eden-gutter').css("transform", "translate(50%,50%) scale(2.0,2.0)"); }}$;

**Comment by ElizabethHudnott** [comment]
What wouldn't you want scaled?  Integrating the buttons into the UI cleanly is also a concern.  I think this requires a face to face discussion to understand the requirements better.

When I tried doing a CSS scale transform it distorted some things in my browser (Chrome), text I think, or possibly form controls.  Would be cleaner (and visually better it seems) to clean up the CSS to represent things using relative font sizes and then just adjust the font size of the appropriate container.

**Comment by knicos** [comment]
agreed about face-to-face. had no visual problems with scale myself? (as long as you scale and translate the correct things). Its A LOT of work just now to make everything font size relative... thinking the syntax highlighter here. I attempted a font-size change in the highlighter and besides things breaking a little it was also very very slow in Chrome for some reason... I think it has real issues rendering thousands of different span elements at large font. On the other hand CSS scale seems as fast as normal.

Was thinking of making a script to append buttons to UI...

import plugins/scriptview/font;

**Comment by ElizabethHudnott** [comment]
Sorry, I can't remember what the issues were, and I was probably enlarging to 500% or something before it became a serious problem, and browsers improve all the time also, but I remember testing it when I implemented canvas scaling and judging the result as inadequate.

**Comment by knicos** [comment]
perhaps its different for canvas, could very well cause font distortion there I think, because fonts are rendered entirely differently.

**Comment by ElizabethHudnott** [comment]
No, this was the HTML parts of the JS-EDEN, not the HTML <canvas> element, so same rendering, but anyway.

**Comment by knicos** [comment]
Done for scriptview. Font buttons can be added as required with the following:

`
import plugins/scriptview/font;
do addFontButtons "input";
`

Where "input" is the view name. Currently the font + and - buttons only go between regular and double size.

**Comment by ElizabethHudnott** [comment]
Just a quick note to add.  The canvas has already had this feature for some time.  Click anywhere in the canvas and then press Alt+, Alt- or Alt0, or (if your touchpad supports it) use the standard pinch to zoom finger gesture.

**Comment by meurigbeynon** [comment]
Definitely something I shall need to see demo-ed face-to-face. It seems 
like I've only just got enough favourably disposed fingers to try this 
out, and so far I have had no success. Plus of coure my touchpad may not 
support it.

Meurig

**Comment by ElizabethHudnott** [comment]
You can tell if your touchpad supports pinch zoom by making the gesture while the mouse pointer is *not* positioned over the canvas, in which case it acts as if you had pressed Ctrl+ or Ctrl-, which is built into the browser and doesn't require any programming.  If your machine doesn't have Windows 8 or later then I doubt that your touchpad supports it.  Note that whole page zoom using the touchpad moves in fixed jumps (just like Ctrl+ and Ctrl-), so it is rather fiddly to make the gesture.  The canvas zoom also goes fixed jumps in Chrome but it is nice and smooth in Firefox and more like the experience tablet users are familiar with, though I haven't tested it on a device with a touch screen, only touchpads.  If it's too fiddly or not possible because of lack of hardware support or browser inadequacies then the keyboard alternatives (Alt+ and Alt-) are rather useful too.  Also, altering the value of _view_picture_zoom will change the zoom factor too.  Earlier today I experimented with making a (meta-)construal which included a JS-EDEN button that altered the zoom, with partial success.

You can also check in your mouse control panel, if there is a page called something like clickpad settings then it is possible that your touchpad supports gestures but they've been explicitly disabled.

My conclusion (from my experiments while constructing this feature) is that pinch to zoom is the best solution in the cases where (a) the hardware supports it (touch screen would be even better because of the direct physical correspondence which is noticeably lacking and adding to the fiddliness of a touchpad), (b) the browser supports it, and (c) the user doesn't suffer with finger joint problems, etc.  Ideally I'd rather avoid cluttering up the UI with clunky buttons that are rarely needed when working on one's own and that don't have a natural place to put them in the UI, as most views don't have existing toolbars.  On the other hand, conditions (a), (b) and (c) are far from always being satisfied(!), though of course things will inevitably get better over time as people upgrade their machines.

**Comment by knicos** [comment]
Just added Alt+, Alt- and Alt0 for zooming in script view. Also added an ..._zoom observable. It can zoom smoothly between any values. Can still add font buttons as before. I'll borrow the canvas pinch code at some point ...? I'll try and add the same zoom code to symbol lists and see what happens.

**Comment by ElizabethHudnott** [comment]
If you prefer not to have the scrollbars when the canvas is zoomed in then you could try something like:

_view_picture_canvas_right is _view_picture_width / _view_picture_zoom;
_view_picture_canvas_bottom is _view_picture_height / _view_picture_zoom - 1;

This has three caveats:

(1) It won't work if the canvas contains HTML elements near to the right or bottom sides because currently the HTML material isn't clipped, only the canvas.  (bug).
(2) I have no idea why the -1 is needed.  Flippin' jQuery Dialog weirdness causing obscure bugs.
(3) If you do this then you'll be restricted to seeing the top-left corner of the canvas and won't be able to pan around to see different parts anymore without the scroll bars.  You could programmatically adjust _view_picture_offset if you want to implement your own customize panning mechanism of some kind.

**Comment by ElizabethHudnott** [comment]
Something like the following is also kind of interesting:

canvasStartingWidth = _view_picture_width;
canvasStartingHeight = _view_picture_height;
_view_picture_width is canvasStartingWidth * _view_picture_zoom;
_view_picture_height is canvasStartingHeight * _view_picture_zoom;
